keps/sig-apps/20190305-add-progress-deadline-seconds-to-job.md

```diff
@@ -0,0 +1,132 @@
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+         * [Detailed Design](#detailed-design)
+            * [Feature Gate](#feature-gate)
+            * [API Object](#api-object)
+         * [Implementation Details/Notes/Constraints](#implementation-details/notes/constraints)
+            * [Implementation Details](#Implementation Details)
+            * [Validation](#Validation)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+
+## Summary
+
+We propose to add an optinal field `ProgressDeadlineSeconds` to `job`. Like `deployment`'s `ProgressDeadlineSeconds`, if this field is set, it is used to fail a `job` if the `job` does not start within the given time, i.e, the progress deadline.
+
+## Motivation
+
+In [#48075][] and [#51153][] we've introduced backoff policy for job, which does not cover certain issues with erroneous jobs. 
+For example, we create a job with miss-typed pull spec, and will find out about it only when the job controller kicks a pod and the pod will result in ImagePullBackOff state. The job will remain active.
+Users would like to fail the job instead of having the job remain active if there's a typo, config error, not-pullable image etc.
+For detailed examples, please see the below user stories.
+
+[#48075]: https://github.com/kubernetes/kubernetes/issues/48075
+[#51153]: https://github.com/kubernetes/kubernetes/issues/51153
+
+### Goals
+
+We'd like to fail a job if it does not start within given time.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### User Stories
+
+* [#67828][] `Jobs with a container config error will never complete.` A job with a config error (pod: Warning Failed 22m (x8 over 24m) kubelet, minikube Error: container has runAsNonRoot and image will run as root), will never complete, the job will not have any success or fail status, and will have an 'activate' forever. While this should count as a failure, and take into account restartPolicy and backoffLimit.
+* [#62816][] `Request for job failure on pod failure rather than job remaining active.` 
+I run a job with a volume that references a non-existent config map and get an error message like "MountVolume.SetUp failed for volume "some-volume" : configmaps "some-missing-config-map" not found". I expect the status returned by an API call to have "failed": 1.
+Would also like the job to fail if the container image can't be pulled. Would be nice to be able to specify in the job spec how many image pull attempts to make before failing. If I have a typo in the image name, or it's not accessible in my Docker registry, I want the job to fail so I can find that out immediately at job submission/creation time, without having to occasionally look through the job list in the dashboard to find ones with the failures recorded in their details.
+
```

Comment:
Please clean up the user stories a little bit. Right now they look like issue reports. Change it to user stories of how users use/benefit from this new feature.